rattan-core/src/cells/token_bucket.rs

```diff
@@ -0,0 +1,968 @@
+                        break;
+                    }
+                }
+            }
+
+            // wait for packet to arrive and check it
+            while self.packet_queue.is_empty() {
+                // the queue is empty, wait for packets to arrive
+                tokio::select! {
+                    biased;
+                    Some(config) = self.config_rx.recv() => {
+                        self.set_config(config);
+                    }
+                    recv_packet = self.egress.recv() => {
+                        match recv_packet {
+                            Some(new_packet) => {
+                                match self.state.load(std::sync::atomic::Ordering::Acquire) {
+                                    0 => {
+                                        return None;
+                                    }
+                                    1 => {
+                                        return Some(new_packet);
+                                    }
+                                    _ => {
+                                        // check the size of packet and drop or enter the queue
+                                        let packet_size = new_packet.l3_length() + self.packet_queue.get_extra_length();
+                                        if packet_size <= self.max_size {
+                                            // enqueue the packet
+                                            self.packet_queue.enqueue(new_packet);
+                                        }
+                                    }
+                                }
+                            }
+                            None => {
+                                // channel closed
+                                return None;
+                            }
+                        }
+                    }
+                }
+            }
+            // packet arrived, check the size of packet and number of tokens
+
+            // get number of tokens added from last_check
+            let now = Instant::now();
+            let buffer_ns =
+                length_to_time(self.token_bucket.buffer, self.token_bucket.rate).as_nanos() as i64;
+            let mtu_ns = length_to_time(self.peak_token_bucket.buffer, self.peak_token_bucket.rate)
+                .as_nanos() as i64;
+
+            let mut toks = min(buffer_ns, (now - self.last_check).as_nanos() as i64);
+
+            let mut ptoks: i64 = min(mtu_ns, toks + self.peak_token_bucket.tokens);
+
+            ptoks -= length_to_time(
+                self.packet_queue.get_front_size().unwrap(),
+                self.peak_token_bucket.rate,
+            )
+            .as_nanos() as i64;
+
+            toks = min(buffer_ns, toks + self.token_bucket.tokens);
+
+            toks -= length_to_time(
+                self.packet_queue.get_front_size().unwrap(),
+                self.token_bucket.rate,
+            )
+            .as_nanos() as i64;
```

Comment:
Actually I have to use negative values in toks and ptoks, because I need to calculate `next_check` according to negative toks and ptoks ( as what is done in tc_tbf )
Also, since I need to modify toks and ptoks for many times in a specific sequence, if I directly modify the value on self.tb.toks or ptoks, the code will be so dirty and unreadable ( at least I think so ).

Reply:
As I want to use `Duration` for toks to avoid such dirty type conversions, which are not allowed for negative values, we could change the logic from another direction. We should decide when we are likely to have the possibility to send packets (by `next_available`). At that point, we should add all available tokens to self.toks. This can be done by such `self.token_bucket.tokens = min(buffer_ns, self.token_bucket.tokens + now - self.last_check)`. And we can first check if we have enough tokens before minus. If we don't have enough tokens, we should 'move' `next_available` to the calculated possible timing and skip the following process.